### PR TITLE
Danbar/e2e rolling update 10

### PR DESCRIPTION
**Pull request by danbar2**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a API change?

```release-note

```

#### Additional documentation e.g., enhancement proposals, usage docs, etc.:

```docs

```

### CODE REVIEW

File: operator/e2e/setup/shared_cluster.go [hunk 1]
```diff
@@ -586,9 +609,11 @@ func setupRegistryTestImages(registryPort string, images []string) error {
 		if err != nil {
 			// Image doesn't exist locally, need to pull it
 			pullOpts := image.PullOptions{}
```
Comment (shayasoolin) [inline operator/e2e/setup/shared_cluster.go]: This section is duplicate with operator/e2e/setup/k8s_clusters.go:prepullImages, please consider extracting to a commonly used func.

**Review by shayasoolin — DISMISSED** [review]

File: operator/e2e/setup/shared_cluster.go [hunk 2]
```diff
@@ -495,6 +507,117 @@ func (scm *SharedClusterManager) Teardown() {
...
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal auth config: %w", err)
+	}
+
+	return base64.StdEncoding.EncodeToString(encodedJSON), nil
+}
+
+// getDockerAuthForImage returns the base64-encoded auth string for pulling an image
+// by reading credentials from the Docker config file.
+func getDockerAuthForImage(imageName string, registryURL string, configPath string) (string, error) {
```
Comment (gflarity) [inline operator/e2e/setup/shared_cluster.go]: Hmm... What image is requiring Docker Auth? I recently replaced the registry image we were using. Ideally we wouldn't get any images from docker hub (the paid requirement to avoid rate limits is annoying) at all. 

If we can just replace the image instead of worry about auth, I'd prefer it. It's not clear the auth would work in CI either?

File: operator/e2e/setup/k8s_clusters.go [hunk 3]
```diff
@@ -777,22 +777,13 @@ func prepullImages(ctx context.Context, images []string, registryPort string, lo
...
-				_, err = io.Copy(logger.WriterLevel(utils.DebugLevel), pullReader)
-			} else {
-				_, err = io.Copy(io.Discard, pullReader)
+				outputWriter = logger.WriterLevel(utils.DebugLevel)
 			}
-			pullReader.Close()
-			if err != nil {
-				errChan <- fmt.Errorf("failed to read pull output for %s: %w", img, err)
+			if err := pullImageIfNotExists(ctx, dockerClient, img, outputWriter); err != nil {
+				errChan <- err
```
Comment (gflarity) [inline operator/e2e/setup/k8s_clusters.go]: We should log if pullImageIfNotExists fails... 


```suggestion
			if err := pullImageIfNotExists(ctx, dockerClient, img, outputWriter); err != nil {
				logger.Debugf("Pull failed for %s: %v", img, err)
				errChan <- err
```

File: operator/e2e/setup/shared_cluster.go [hunk 4]
```diff
@@ -495,6 +507,117 @@ func (scm *SharedClusterManager) Teardown() {
 	}
 }
 
+// parseRegistryFromImage extracts the registry host from an image name.
+// Returns the provided defaultRegistry if the image doesn't specify a custom registry.
+func parseRegistryFromImage(imageName string, defaultRegistry string) string {
```
Comment (gflarity) [inline operator/e2e/setup/shared_cluster.go]: We can replace this manual string parsing with the distribution/reference package, which is already an indirect dependency in our go.mod. It handles OCI edge cases and implicit Docker Hub domains (like index.docker.io) much more reliably than strings.SplitN.

```go
named, _ := reference.ParseNormalizedNamed(imageName)
registryHost := reference.Domain(named)
```

File: operator/e2e/setup/shared_cluster.go (hunk 2 above)
Comment (gflarity) [reply operator/e2e/setup/shared_cluster.go]: Also, if we really do need auth, the current manual JSON parsing of .docker/config.json will fail for developers using credential helpers (e.g., osxkeychain, pass, or ecr-login), as the password won't be in the JSON file. Since we already have github.com/docker/cli in our tree, we should use config.LoadDefaultConfigFile(os.Stderr) to handle auth natively.

```go
// Suggestion: Leverage Docker's own config loader to support credential helpers
cfg := config.LoadDefaultConfigFile(os.Stderr)
authConfig, err := cfg.GetAuthConfig(registryHost)
```

File: operator/e2e/setup/shared_cluster.go [hunk 5]
```diff
@@ -495,6 +507,117 @@ func (scm *SharedClusterManager) Teardown() {
...
+		ioutil.CloseQuietly(pullReader)
+		if pullErr != nil {
+			return fmt.Errorf("failed to read pull output for %s: %w", imageName, pullErr)
+		}
+	}
+	return nil
+}
+
 // setupRegistryTestImages sets up test images in the registry
 func setupRegistryTestImages(registryPort string, images []string) error {
```
Comment (gflarity) [inline operator/e2e/setup/shared_cluster.go]: Looks like I should have just used google/go-containerregistry as well here to significantly simplify this entire flow. Looks like there's a package called 'crane'  that would allow us to replace the manual Pull → Tag → Push logic (and all the associated auth boilerplate) with a single crane.Copy() call.

File: operator/internal/controller/podclique/reconcilestatus.go [hunk 6]
```diff
@@ -124,10 +124,15 @@ func mutateReplicas(pclq *grovecorev1alpha1.PodClique, podCategories map[corev1.
...
 	// else take it from the PodClique.Status.CurrentPodTemplateHash field
 	if componentutils.IsPCLQUpdateInProgress(pclq) {
 		expectedPodTemplateHash = pclq.Status.RollingUpdateProgress.PodTemplateHash
+	} else if componentutils.IsLastPCLQUpdateCompleted(pclq) {
+		// Rolling update just completed (UpdateEndedAt is set) - use the new hash from RollingUpdateProgress.
+		// This handles the window between markRollingUpdateEnd() and mutateCurrentHashes() where
+		// CurrentPodTemplateHash hasn't been updated yet but the rolling update has finished.
+		expectedPodTemplateHash = pclq.Status.RollingUpdateProgress.PodTemplateHash
 	} else if pclq.Status.CurrentPodTemplateHash != nil {
 		// CurrentPodTemplateHash will be set if RollingUpdateProgress is nil or if the last update has completed.
```
Comment (gflarity) [inline operator/internal/controller/podclique/reconcilestatus.go]: This comment is now stale with change above:

```suggestion
		// Steady state: no rolling update has ever been tracked for this PodClique.
		// This handles fresh PodCliques that have been reconciled but never updated.
```

File: operator/e2e/setup/shared_cluster.go (hunk 5 above)
Comment (danbar2) [reply operator/e2e/setup/shared_cluster.go]: decided to revert the e2e setup changes, @Ronkahn21 raised concerns about it too and it does not related to the bug fix. Will do in future PR

File: operator/e2e/setup/shared_cluster.go (hunk 4 above)
Comment (danbar2) [reply operator/e2e/setup/shared_cluster.go]: decided to revert the e2e setup changes, @Ronkahn21 raised concerns about it too and it does not related to the bug fix. Will do in future PR

File: operator/e2e/setup/k8s_clusters.go (hunk 3 above)
Comment (danbar2) [reply operator/e2e/setup/k8s_clusters.go]: decided to revert the e2e setup changes, @Ronkahn21 raised concerns about it too and it does not related to the bug fix. Will do in future PR

File: operator/e2e/setup/shared_cluster.go (hunk 2 above)
Comment (danbar2) [reply operator/e2e/setup/shared_cluster.go]: decided to revert the e2e setup changes, @Ronkahn21 raised concerns about it too and it does not related to the bug fix. Will do in future PR

File: operator/e2e/setup/shared_cluster.go [hunk 7]
```diff
@@ -495,6 +507,117 @@ func (scm *SharedClusterManager) Teardown() {
...
+	// Check if image already exists locally
+	_, _, err := cli.ImageInspectWithRaw(ctx, imageName)
+	if err != nil {
+		// Image doesn't exist locally, need to pull it
+		pullOpts := image.PullOptions{}
+		authStr, authErr := getDockerAuthForImage(imageName, defaultDockerRegistry, defaultDockerConfigPath)
+		if authErr == nil {
+			pullOpts.RegistryAuth = authStr
+		}
+		// Ignore auth errors and attempt pull without auth - may work for public images
```
[the remark this answers is not in the document]
Comment (danbar2) [reply operator/e2e/setup/shared_cluster.go]: decided to revert the e2e setup changes

File: operator/e2e/tests/rolling_updates_test.go [hunk 8]
```diff
@@ -44,8 +49,10 @@ func Test_RU7_RollingUpdatePCSPodClique(t *testing.T) {
 		t.Fatalf("Failed to update PodClique spec: %v", err)
 	}
 
+	// With MaxSurge rolling update strategy, each pod requires: create surge pod -> wait Ready -> delete old pod
+	// This takes longer than delete-first approach, so we use a longer timeout
 	tcLongTimeout := tc
-	tcLongTimeout.Timeout = 1 * time.Minute
+	tcLongTimeout.Timeout = 5 * time.Minute
```
Comment (gflarity) [inline operator/e2e/tests/rolling_updates_test.go]: Are we sure 5 minutes is required?  Seems very long.

File: operator/internal/controller/podclique/reconcilestatus.go [hunk 9]
```diff
@@ -124,12 +124,13 @@ func mutateReplicas(pclq *grovecorev1alpha1.PodClique, podCategories map[corev1.
 // mutateUpdatedReplica calculates and sets the number of pods with the expected template hash
 func mutateUpdatedReplica(pclq *grovecorev1alpha1.PodClique, existingPods []*corev1.Pod) {
 	var expectedPodTemplateHash string
-	// If the PCLQ update is in progress then take the expected PodTemplateHash from the PodClique.Status.RollingUpdateProgress.PodCliqueSetGenerationHash field
-	// else take it from the PodClique.Status.CurrentPodTemplateHash field
-	if componentutils.IsPCLQUpdateInProgress(pclq) {
+	// If RollingUpdateProgress exists (update in progress or recently completed), use the target hash from it.
+	// This covers both the active update phase and the window after completion before CurrentPodTemplateHash is synced.
+	if pclq.Status.RollingUpdateProgress != nil {
```
Comment (gflarity) [inline operator/internal/controller/podclique/reconcilestatus.go:129]: Consider adding a unit test for `mutateUpdatedReplica` that specifically covers this "just-completed" state (where `RollingUpdateProgress != nil` and `UpdateEndedAt != nil`). The existing e2e tests validate the rolling update completes, but a focused unit test would help prevent regressions and document this edge case more explicitly.

Something like:
```go
{
    name: "update just completed - use RollingUpdateProgress hash",
    pclq: &PodClique{
        Status: PodCliqueStatus{
            RollingUpdateProgress: &RollingUpdateProgress{
                PodTemplateHash: "new-hash",
                UpdateEndedAt:   &metav1.Time{Time: time.Now()},
            },
            CurrentPodTemplateHash: ptr.To("old-hash"),
        },
    },
    // ... pods with "new-hash" should be counted as updated
}
```

**Review by gflarity — COMMENTED** [review]
Looking good, one question and one comment.

File: operator/e2e/tests/rolling_updates_test.go (hunk 8 above)
Comment (danbar2) [reply operator/e2e/tests/rolling_updates_test.go]: it actually ends in 30s, so we can keep it with 1m if we hit timeout I suggest to increase that
